src/signature/spki.rs

```diff
@@ -0,0 +1,488 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHORS DISCLAIM ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR
+// ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+use der;
+use signature;
+use untrusted;
+
+/// An error that occurs during certificate validation or name validation.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum VerifyWithSPKIError {
+    /// The encoding of some ASN.1 DER-encoded item is invalid.
+    BadDER,
+
+    /// The signature is invalid for the given public key.
+    InvalidSignatureForPublicKey,
+
+    /// The SignatureAlgorithm does not match the algorithm of the SPKI.
+    /// A mismatch could be because of the algorithm (RSA vs DSA, etc) or the
+    /// parameters (ECDSA_p256 vs ECDSA_384, etc).
+    UnsupportedSignatureAlgorithmForPublicKey,
+
+}
+
+pub fn verify_signature(signature_alg: &SignatureAlgorithm,
+                        spki_value: untrusted::Input, msg: untrusted::Input,
+                        signature: untrusted::Input) -> Result<(), VerifyWithSPKIError> {
+    let spki = try!(parse_spki_value(spki_value));
+    if !signature_alg.public_key_alg_id
+        .matches_algorithm_id_value(spki.algorithm_id_value) {
+        return Err(VerifyWithSPKIError::UnsupportedSignatureAlgorithmForPublicKey);
+    }
+    signature::verify(signature_alg.verification_alg, spki.key_value, msg,
+                      signature)
+        .map_err(|_| VerifyWithSPKIError::InvalidSignatureForPublicKey)
+}
+
+struct SubjectPublicKeyInfo<'a> {
+    algorithm_id_value: untrusted::Input<'a>,
+    key_value: untrusted::Input<'a>,
+}
+
+// Parse the public key into an algorithm OID, an optional curve OID, and the
+// key value. The caller needs to check whether these match the
+// `PublicKeyAlgorithm` for the `SignatureAlgorithm` that is matched when
+// parsing the signature.
+fn parse_spki_value(input: untrusted::Input)
+                    -> Result<SubjectPublicKeyInfo, VerifyWithSPKIError> {
+    input.read_all(VerifyWithSPKIError::BadDER, |input| {
+        let algorithm_id_value =
+        try!(der::expect_tag_and_get_value(input, der::Tag::Sequence)
+            .map_err(|_| VerifyWithSPKIError::BadDER));
+        let key_value = try!(der::bit_string_with_no_unused_bits(input)
+            .map_err(|_| VerifyWithSPKIError::BadDER));
+        Ok(SubjectPublicKeyInfo {
+            algorithm_id_value: algorithm_id_value,
+            key_value: key_value,
+        })
+    })
+}
+
+
+/// A signature algorithm.
+pub struct SignatureAlgorithm {
+    public_key_alg_id: AlgorithmIdentifier,
+    signature_alg_id: AlgorithmIdentifier,
```

Comment:
It might be a good idea to rename `SignatureAlgorithm` to `AlgorithmIdentifier`, leaving the `Signature` part implied by the fact that it is nested in `ring::signature`.